src/Hypsography/Hypsography.jl

```diff
@@ -64,180 +122,156 @@ If the decay-scale is poorly specified (i.e., `s * zₜ` is lower than the maxim
+    if η <= ηₕ
+        return Geometry.ZPoint(
+            η * z_top.z +
+            z_surface.z * (sinh((ηₕ - η) / s / ηₕ)) / (sinh(1 / s)),
+        )
+    else
+        return Geometry.ZPoint(η * z_top.z)
+    end
 end
 
-# deprecated in 0.10.12
-LinearAdaption() = LinearAdaption(nothing)
-
-@deprecate(
-    ExtrudedFiniteDifferenceSpace(
-        horizontal_space::Spaces.AbstractSpace,
-        vertical_space::Spaces.FiniteDifferenceSpace,
-        adaption::LinearAdaption{Nothing},
-        z_surface::Fields.Field,
-    ),
-    ExtrudedFiniteDifferenceSpace(
-        horizontal_space,
-        vertical_space,
-        LinearAdaption(z_surface),
-    ),
-    false
+function physical_z_to_ref_z(
+    adaption::SLEVEAdaption,
+    z_physical::Geometry.ZPoint,
+    z_surface::Geometry.ZPoint,
+    z_top::Geometry.ZPoint,
 )
+    error("This method is not implemented")
+end
 
-# linear coordinates
+# can redefine this constructor for e.g. multi-arg SLEVE
 function _ExtrudedFiniteDifferenceGrid(
     horizontal_grid::Grids.AbstractGrid,
     vertical_grid::Grids.FiniteDifferenceGrid,
     adaption::HypsographyAdaption,
+    global_geometry::Geometry.AbstractGlobalGeometry,
 )
-    if adaption isa LinearAdaption
-        if isnothing(adaption.surface)
-            error("LinearAdaption requires a Field argument")
-        end
-        if axes(adaption.surface).grid !== horizontal_grid
-            error("Terrain must be defined on the horizontal space")
-        end
-    end
+    @assert Spaces.grid(axes(adaption.surface)) == horizontal_grid
+    z_surface = Fields.field_values(adaption.surface)
+
+    face_z_ref =
+        Grids.local_geometry_data(vertical_grid, Grids.CellFace()).coordinates
+    vertical_domain = Topologies.domain(vertical_grid)
+    z_top = vertical_domain.coord_max
 
-    # construct initial flat space, then mutate
-    ref_grid = Grids.ExtrudedFiniteDifferenceGrid(
+    face_z =
+        ref_z_to_physical_z.(Ref(adaption), face_z_ref, z_surface, Ref(z_top))
+
+    return _ExtrudedFiniteDifferenceGrid(
         horizontal_grid,
         vertical_grid,
-        Flat(),
+        adaption,
+        global_geometry,
+        face_z,
     )
-    face_ref_space = Spaces.FaceExtrudedFiniteDifferenceSpace(ref_grid)
-    face_ref_coords = Spaces.coordinates_data(face_ref_space)
-    coord_type = eltype(face_ref_coords)
-    z_ref = face_ref_coords.z
-
-    vertical_domain = Topologies.domain(vertical_grid)
-    z_top = vertical_domain.coord_max.z
-
-    grad = Operators.Gradient()
-    z_surface = Fields.field_values(adaption.surface)
+end
 
-    FT = eltype(z_surface)
+# generic 5-arg hypsography constructor, uses computed face_z points
+function _ExtrudedFiniteDifferenceGrid(
+    horizontal_grid::Grids.AbstractGrid,
+    vertical_grid::Grids.FiniteDifferenceGrid,
+    adaption::HypsographyAdaption,
+    global_geometry::Geometry.AbstractGlobalGeometry,
+    face_z::DataLayouts.AbstractData{Geometry.ZPoint{FT}},
+) where {FT}
```

Comment:
Is this only expected to work with `ZPoint`s? I'm wondering if we need this specialization.

Reply:
I think so, the adaption function only work with `ZPoints`, so `face_z` has to be `ZPoint` as well.

Reply:
Why is it not valid for `XYZPoint`s? (I have no objection, I'm just curious)

Reply:
I don't really know. I think this function is extruding an horizontal grid using `face_z`, so the only value that makes sense is `Z`?

Reply:
Ok, I'm fine with this, then. Extending support for more types is easier than removing them (since removing is breaking)